### PR TITLE
SVGD implementation

### PR DESCRIPTION
**Pull request by antotocar34**
As discussed in #385.

Done:
- Implementation of the algorithm
- Added very simple (bad) test.

TODO:
- Add an example with GPU where gradients is calculated with pmap.
- How to take mini batch gradients? 
- Come up with a better test.

**Appreciate any comments & feedback!**

 A few important guidelines and requirements before we can merge your PR:

 - [ X ] *If I add a new sampler*, there is an issue discussing it already;
 - [ X ] We should be able to understand what the PR does from its title only;
 - [ X ] There is a high-level description of the changes;
 - [ X ] There are links to *all* the relevant issues, discussions and PRs;
 - [ X ] The branch is rebased on the latest `main` commit;
 - [ ] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [ ] The code respects the current naming conventions;
 - [ ] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [ X ] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [ ~ ] There are tests covering the changes;
 - [ ] The doc is up-to-date;
 - [ ] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

### CODE REVIEW

**Comment by antotocar34** [comment]
I've made a quick collab notebook to play around with the algorithm.

https://colab.research.google.com/drive/115rjkthFjz10d_l_M2BqAQdn-kaDdulC?usp=sharing

File: blackjax/vi/svgd.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    Parameters
+    ----------
+    initial_particles
+        Initial set of particles to start the optimization
+    kernel_paremeters
+        Arguments to the kernel function
+    optimizer
+        Optax compatible optimizer, which conforms to the `Optimizer` protocol
+    """
+    particle_array, _ = particles_to_array(initial_particles)
```
Comment (albcab) [inline blackjax/vi/svgd.py:46]: you might need to extract the number of particles once here and include it in `SVGDState`

**Review by albcab — CHANGES_REQUESTED** [review]
Thanks for this PR!

Let me know if there is anything that needs clarification on my review (or if I've missed anything).

To be able to do pmap and to be able to use minbatches for the gradient computations, you can consider taking as input an optional `grad_estimator` or `grad_fn` that allows the user to specify its own gradient computations, and have it default to the usual `jax.grad` of the log density. In the case of minibatches, the script `sgmcmc/gradients.py` has all the basic components you'll need already implemented!

An example notebook meant to run on GPU can be left as a separate PR for `blackjax/sampling-book`, it would make a very nice addition to this implementation. 

I might be the best use of your time to leave the tests for when you are on the final draft of the implementation, i.e. when you have minibatch gradients, pmapping, etc.

**Comment by antotocar34** [comment]
Thanks for the detailed comments! I'll try and get to them next week :smiley:

**Comment by albcab** [comment]
I had to implement SVGD for a project I'm working on. The latest commit is a working version with the improvements from my review. I've tested it on medium-sized models and the vmapping hasn't broken yet, so we can leave it like this for now. It might still need some work on the documentation, other than that it should be ready to merge.

**Review by albcab — DISMISSED** [review]

**Review by albcab — APPROVED** [review]

**Comment by junpenglao** [comment]
Thank you both!
